File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    static int[] dy = {1, -1, 0, 0};
    static int[] dx = {0, 0, 1, -1};
    
    public int solution(int[][] maps) {
        
        int row = maps.length;
        int col = maps[0].length;
        
        boolean[][] visited = new boolean[row][col];
        
        return bfs(row, col, maps, visited);
    }
    
    static int bfs(int row, int col, int[][] maps, boolean[][] visited) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{0, 0, 1});
        
        visited[0][0] = true;
        
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int y = cur[0];
            int x = cur[1];
            int cnt = cur[2];
            
            if(y == row - 1 && x == col - 1) {
                return cnt;
            }
            
            for(int i = 0; i < 4; i++) {
                int ny = y + dy[i];
                int nx = x + dx[i];
                
                if(ny >= row || ny < 0 || nx >= col || nx < 0) continue;
                if(visited[ny][nx]) continue;
                if(maps[ny][nx] != 1) continue;
                
                q.add(new int[]{ny, nx, cnt + 1});
                visited[ny][nx] = true;   
            }
        }
        
        return -1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])

    visited = [[False] * m for _ in range(n)]
    q = deque()

    # y, x, dist
    q.append((0, 0, 1))
    visited[0][0] = True

    dy = [1, -1, 0, 0]
    dx = [0, 0, 1, -1]
    
    while q:
        y, x, dist = q.popleft()

        if y == n - 1 and x == m - 1:
            return dist

        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]

            if ny >= len(maps) or ny < 0 or nx >= len(maps[0]) or nx < 0:
                continue
            if maps[ny][nx] == 0:
                continue
            if visited[ny][nx]:
                continue
            
            visited[ny][nx] = True
            q.append((ny, nx, dist + 1))

    return -1